### DISCUSSION TITLE
@xstate/angular?

### DISCUSSION THREAD

**Discussion by keithernet** [statelyai/xstate#3205] [Ideas]
Kicking off a discussion about what an @xstate/angular package would look like. My initial questions revolve around the patterns that a "blessed" library should encourage or discourage? Is it possible to build a class or a helper that makes it as simple as `@xstate/React`'s useMachine function to integrate a machine with a component?

**Comment by keithernet** [comment]
[Here's a sample repository](https://github.com/stefanoslig/angular-xstate/tree/master/src/app/auth/%2Bxstate) I found to give an overview of what using XState in Angular can look like.

**Reply by Avisa-GA** [reply]
How we can generate these files?

**Reply by keithernet** [reply]
I'm not sure what you mean. Which files are you hoping to generate?

**Reply by Avisa-GA** [reply]
I got my answer from David, Thank You.

**Reply by SerkanSipahi** [reply]
I think with generate it means to generate files automated (schematics) by the angular-cli, right?

**Comment by Andarist** [comment]
I don't think that any of us have enough experience with Angular to know that ecosystem's conventions. We would be open to publishing a package like this - but it would be best if the community member could create it in this repo.

**Reply by keithernet** [reply]
My company will be embarking on a project to convert our NgRx code to XState and will likely be building a library of some sort. I can't commit my company to supporting the effort indefinitely, but we would be open to contributing the initial code and fixing issues we encounter along the way for sure. What would the process for starting up a new package look like in that regard?

**Reply by davidkpiano** [reply]
We'll write contributing guidelines for this 👍

**Comment by adevil5** [comment]
I've got some code that I never quite got working how I wanted. But basically I made useMachine, useInterpret, useSelector, and useActor functions that return RxJS observables. 

I also have a StateMachineServiceBase class that wraps the useMachine function, making it easily extendable by an Angular service.

But I think it's missing something, something to make it that last bit easy to get either a global or scoped Angular service that automatically stops the XState interpretor in the OnDestroy lifecycle hook. 

I'll upload it to GitHub and link it in case you want to peek at it.

**Reply by davidkpiano** [reply]
Yes please!

**Reply by adevil5** [reply]
Here's the repo: https://github.com/adevil5/xstate-libs

I have two packages in there. @xstate-libs/rxjs and @xstate-libs/angular. I figure it's a good path to get a good RxJS layer, and then we could add in Angular-specific goodies as needed on top. 

Starting with the "packages\xstate-rxjs\src\test\use-machine.spec.ts" tests might be the easiest way to see the core of what xstate/rxjs can do. I created a useMachine function that works similarly to the other XState libraries. It can return an observable of the state (state$), a send function (send), the interpreted service instance (service), and a select function (select), which is my attempt at an NgRx selector-equivalent. I also tried to implement some of the same tests that the xstate/react library has, but I got lazy.

Here's a snippet of the useMachine test file.
```ts
describe('useMachine', () => {
  const xtc: XstateTestCleaner = new XstateTestCleaner();

  afterEach(() => {
    xtc.clean();
    jest.useRealTimers();
  });

  const fetcher = (
    onFetch: () => Promise<any> = () => Promise.resolve('some data'),
    persistedState?: AnyState
  ) => {
    const { state$, send, service } = useMachine(fetchMachine, {
      services: {
        fetchData: onFetch,
      },
      state: persistedState,
    });

    xtc.addXstateService(service);

    return { state$, send, service };
  };

  test('should work with the useMachine hook', (done) => {
    const { state$, send } = fetcher(() => Promise.resolve('fake data'));

    state$.subscribe((state) => {
      if (state.context.data === 'fake data') {
        done();
      }
    });

    send({ type: 'FETCH' });
  });

  test('should work with the useMachine hook (rehydrated state)', (done) => {
    const { state$ } = fetcher(
      () => Promise.resolve('fake data'),
      persistedFetchState
    );

    state$.subscribe({
      next: (state) => {
        if (state.context.data === 'persisted data') {
          done();
        }
      },
      error: (err) => done(err),
    });
  });
});
```

The next convenience I added was wrapping the useMachine function in a class. That way you could easily extend an Angular service with it like this:
```ts
const MACHINE_STORAGE_KEY = 'tenant';

@Injectable({
  providedIn: 'root',
})
export class TenantService extends StateMachineServiceBase<
  typeof tenantMachine
> {
  isTenantLoading$: Observable<boolean>;

  constructor(private apiService: ApiService) {
    super(tenantMachine, {
      storageKey: MACHINE_STORAGE_KEY,
      services: {
        fetchTenant: (context) => this.fetchTenant(context.hostName),
      },
      devTools: !environment.production,
    });

    this.isTenantLoading$ = this.select((state) => !state.matches('idle'));
  }

  private async fetchTenant(hostName: string) {
    const response = await this.apiService.client.query({
      tenantByHostName: [
        {
          hostName,
        },
        {
          description: true,
          hostName: true,
          tenantId: true
        },
      ],
    });
    return response.tenantByHostName;
  }
}
```

This repo is very much a work-in-progress. Unfortunately, I had to pivot to other work. If you think it can help here, this code is free for anyone to use and build upon/steal from/etc.

**Reply by keithernet** [reply]
Thanks for sharing! I went down a similar path and found the host service to be a bit cumbersome due the need to call the super constructor before any other code. Having to either define either the entire options object in `super(...)` or in a separate, also injected, class felt like a frustrating constraint to me. Have you found this to be the case in practice or did you not get the chance to try it out before moving on to your other work?

Either way, I'll be sure to check out the code. Thanks again!

**Reply by keithernet** [reply]
One option to work around the super constraint could be using an abstract initialize hook that the extending class needed to implement. [Something like this.](https://www.typescriptlang.org/play?#code/FAQwRgzgLgTiDGUAE8A2IISQIQwUwB4AVAPiQG9gkkBbBACwEsA7PAQgC4kiBuJAen5IA7gHsArqgAmSeiABueJFFFIA5nmRR6SjaxiN4VFKObQY4xKJgAKAJSVq1bYwgA6FoyiMQqRgC88e2MAX2BjcHMEZE9vXwCguy55UUYpHnDqAAdxMD94JAg8VDxEABIkAF4kGyKSqxgueyqyEGYATzsWihCBfilVOtLkaHEAMzHgMOA0DCwAUQAPKDxmKRY1JDxl1aksXCKCcw2ySmN4U3NLFVssg3kQFcLRGjwASWYAK2G8KQBlPAweSGPBcNqdRxOCDiLKA+wZai9YzGWI+PyBeyQ6iCJADWgMFhKUYTWSAvBuUJTcIXMwxKpIVjCJBLFZrDY2ADkEBemiYzDUHLsGWAjDcQ3KQA) That way you could use other methods on the extending class within the configuration of the machine.

**Reply by adevil5** [reply]
Personally, passing the options into a super call didn't bother me. My Angular service constructor bodies are often empty anyway, so the super call wasn't jumbled up with other kinds of code setup. And it's easy enough to think of these as a specific type of Angular service where they almost become identifiable by the big super call.

The core of the work in my repo is the useMachine function. So whatever way you want to call the function and expose its return values should work. So while I'm personally cool with a big super call, you could also refactor the existing base class to be abstract with an abstract initMachine(machine, options) method like you're suggesting. I suppose a third option is just supplying the underlying useMachine function and letting developers hook them into Angular services however they want. 

The only thing I don't know about with your approach is if TypeScript is smart enough to know that the abstract base class's properties (state$, send, service, select) won't be undefined because they're not set statically or directly in the constructor. I think that was a TypeScript issue at some point, and I don't want to initialize those properties to any-ish types to get around that blocker and lose the typing specific to a particular machine. But I'm no typescript magician, so I could have that wrong.

**Reply by adevil5** [reply]
And to answer another question you had in there, I only started trying to use this library in a dev-like environment. But it's not nearly battle-tested enough for production yet. It needs more tests and real-world use.

**Reply by keithernet** [reply]
I am definitely not a typescript magician either so I'd need to play with it to see if it's possible. 

I mostly missed the ability to define a method for a network call or something in an invoke, but I can see your point. 

Thanks for following up and again for sharing the code. It at not be battle ready but it's a great way to start discussion.

**Reply by adevil5** [reply]
> the ability to define a method for a network call or something in an invoke
I’m trying to understand what you mean by this. The sample TenantService angular service I have above has a fetchService method that defines a network call, and it’s passed to the machine to be invoked. I think I’m missing something. 

And no problemo, anything to help out our Angular comrades.

**Reply by keithernet** [reply]
Maybe I making some silly mistake, but I didn't think it was legal to pass a class field into super because they're not accessible yet. [Should this work?](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCkFMDC4rQN4ChPV9YA9gHYQAuATgK7CkHkAUEBAtggGKVE0CWxAXNHoBKaAF4AfNABuBbgBMhWHHgC+mNZlCQYSABbcQclNugIAHqQRE5MeBGSoYWPPmJkqNOsOcu8ESgAOCAyk+hAAdMwAnhxcpLxEQgDcyrgaLgHk3FJgltDRsTzEwhjYLmoqQA)

**Reply by adevil5** [reply]
As far as I'm aware, you're right about not being able to call `this` in the super call. But you can reference `this` in a callback function.
[Your example with one tweak](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCkFMDC4rQN4ChPV9YA9gHYQAuATgK7CkHkAUEBAtggGKVE0CWxAXNHoBKaAF4AfNABuBbgBMhWHHgC+mNZlCQYSABbcQclNugIAHqQRE5MeBGSoYWPPmJkqNOsOcu8ESgAOCAzCYpKk+hAAdMwAnhxcpLxEQgDcyrgaLgHk3FJgltBxCTzEoUouaipAA)

```
services: {
  fetchTenant: (context) => this.fetchTenant(context.hostName),
},
```
This code snippet taken from the larger TenantService example above shows the tenantMachine's service, `fetchTenant`, being set with XState's usual "(context, event) => <code\>" callback pattern.

**Reply by keithernet** [reply]
Ah, that's what I missed! Thanks!

**Comment by adevil5** [comment]
I'd also like to posit a small Angular-ism in naming machine files/folders (just for @xstate/angular users). Angular folks are used to certain file/folder naming/structure conventions. 

The point I feel most strongly about is a file naming convention like this: `<name>.machine.ts`

Additionally, because the VS Code typegen gives us an extra file, the machine file should be in a folder named after the machine, optionally nested in a machines folder if there are multiple machines where you're placing them. 

Example:
- /machines
  - /image-processing
    - image-processing.machine.ts
    - image-processing.machine.typegen.ts

**Reply by keithernet** [reply]
We are currently using a separate file for the machine definition using createMachine (`<name>.machine-def.ts`) and the Angular service that hosts it (`<name>.machine.ts`) and provides implementations. This started, in part, because the VSCode plugin wasn't reliably finding the machines and offering the inspector button when it was living alongside a class however I haven't tested this recently. If that's still an issue we may just be able to ask the XState team to take a look and see if they can resolve that particular issue.

**Reply by adevil5** [reply]
Oh, to be clear I'm just talking about the machine definition file above. If I want to wrap the machine in a global angular service, I'd put it in a separate file in the same folder.

- /machines
  - /image-processing
    - image-processing.machine.ts
    - image-processing.machine.typegen.ts
    - image-processing.service.ts

I didn't want to have a standard that combined the machine and Angular service in one file, in case we wanted to host a machine in a component or directive, to live and die with their lifecycle.

**Reply by keithernet** [reply]
Yep, makes sense!

**Comment by tzkmx** [comment]
In our company xstate is replacing not a structured state management beyond Angular using Rx itself.

The easiest way so far has been vanilla services, in many cases exposing Observables and methods to change that state.

Therefore we are testing approaches like injecting the machine interpreted as the service itself (without Injectable class wrapper), and putting the responsibility of navigation on some global app state machine.

About this approach, the child machine actor sometimes is spawned by the parent machine, but we have a lot of machines scoped to a single component, sending messages to another machines injected in the component and not handled the communication between them by xstate internally.

**Comment by dj-fiorex** [comment]
Hello guys, another Angular developer here, 4 years of working with Angular and loving it! But i'm new XState and want to use it with Angular. what i want to know is: is there someone who are working to a solution and want some help? is there a repo or something? i fall in love with XState and want to try it right now :smile:

**Reply by davidkpiano** [reply]
Yes, we've had an initial discussion with @keithernet on how to do this, so we're getting somewhere! Will keep everyone updated as we make progress.

**Reply by tzkmx** [reply]
> is there a repo or something? i fall in love with XState and want to try it right now 😄

It's not official or anything but a sandbox of experiments around finding patterns to share of using xstate with Angular.

I'm testing right now how to inject directly an interpreted machine onto a component directly as a service using a custom injection token:

https://github.com/tzkmx/levo-ang-sandbox/commit/9d820de10f92309e753c2c051b394e2fc0155ea9#diff-2f28ffa8ceb914787ba21afb9afb09b4897784a5fdb8eea32b27231de8dab346R22

**Reply by stefumies** [reply]
Link is broken, any alternative url?

**Comment by Avisa-GA** [comment]
Same I can’t wait to try xstate with Angular 


Sent from Yahoo Mail for iPhone


On Sunday, June 19, 2022, 2:42 PM, David Khourshid ***@***.***> wrote:




Yes, we've had an initial discussion with @keithernet on how to do this, so we're getting somewhere! Will keep everyone updated as we make progress.

**Comment by keithernet** [comment]
I had a discussion with David and then went on vacation for a week so it may take me a little to get back up to speed, but using XState in angular is totally possible right now and, in fact, works well without any additional adapter library. There's a channel in the Stately Discord where I, and other members of the community, are happy to help.

**Comment by adevil5** [comment]
If you're reading this, it can't hurt to upvote this. https://statelyai.canny.io/xstate/p/xstate-angular

**Comment by lateshift** [comment]
Are there any updates on the Angular support?

Thanks!  🙏

**Reply by keithernet** [reply]
Sorry, not on my end. I shifted away from doing Typescript work at my day job and haven't had a chance to pick up on this. Up the thread adevil5 posted a link to his repository which is think is a really great place to look for inspiration. I'm hoping that I'll have more time to devote to this maybe in March or so, but until then I think that repo is a really great starting point.

**Comment by stefumies** [comment]
We have been waiting for some kind soul to address the blatant lack of any comprehensive Angular Examples integrating XState, especially using Angular's excellent Service architecture and Observables. Its 2023 and there seems to be still no movement on this apart from a scattered collection of custom implementations, most of which throw errors, or don't appear to follow XState conventions at all.

Personally. I'm struggling to produce more than a hack to include a state machine into Angular, nothing of note and certainly not production ready. Now there is a huge project on the horizon and we are tied into Angular, but can offload a huge task requirements with XState. Unfortunately the past initial days have been trying to come up with a secure and prod-ready implementation for XState into the NG ecosystem. 

XState is fantastic, its such a shame that Angular developers are still left hanging

**Reply by Andarist** [reply]
Perhaps you'd like to schedule a call with me to brainstorm how this could be implemented?

**Reply by stefumies** [reply]
> Perhaps you'd like to schedule a call with me to brainstorm how this could be implemented?

Kind offer indeed,  but Im afraid  don't know enough about XState internals to be productive.

 Im approaching this merely as a Developer / User. I would like to see a very simple example (such as the Traffic light example) in Angular. Not with custom hacks and such but showing how to get and read the state in an Angular Service, via Rxjs or not, there need be no async calls, or anything. 

React has useMachine hook to offload this as well as matches etc, it's just strange to me that none of this appears to be approached in Angular despite its widespread usage. As I mentioned before all found examples are usually far too involved overly complex or merely far to abstract to understand fully.

Eventually any support would have to be fully documented with TS types, as there is also very little on this as well.

In brief:

How to get my State Machine into an Angular service, that's the crux.

**Reply by davidkpiano** [reply]
> Im approaching this merely as a Developer / User. I would like to see a very simple example (such as the Traffic light example) in Angular. Not with custom hacks and such but showing how to get and read the state in an Angular Service, via Rxjs or not, there need be no async calls, or anything.

Can you start by showing how you would do this _without_ XState, and we can take that code and translate it to an XState example?

**Comment by stefumies** [comment]
I don't quite follow. How would I do what? The Traffic light example is a widely used example of state control. In fact there are numerous XState examples on it. It's not a problem on doing the Traffic light problem it's a matter of no examples on how to implement the same pattern in Angular,

Without any state control it's a rudimentary switch or similar, it would be nonsensical to create such an app now., as a mere elucidation of my core point. The core point being that , there is no comprehensive guide on any practices (best or not) on how to integrate XState into Angular. There is are reims of info on React, and even Svelte (which has barely left the starting post), but non on Angular which is well past  entrenched now.

Our current project is huge and will involve a substantial state management system, but as I mentioned we are tied to Angular. So for now, it looks like we are going to have to build a custom integration for XState, I wasn't intending to get sidetracked into filling those missing areas in in XState, no matter how interesting that may be (and it would)

The Traffic light example I pulled from memory as it is clear and simple and commands no bloated approach to state management, though it could be anything even a simple Toggle button via Observers / Subjects in Angular. The fundamental issue being how to correctly read from the state machine, and fire events to update via Angular's Rxjs implementation or similar.  No need to understand the creation of the machine itself, that is particularly easy as your Documentation is excellent.

**Reply by with-heart** [reply]
I'll make this easy for you:

No one at Stately knows Angular. 

No one from the Angular community has taken the time to contribute to the docs. 

No one is going to learn Angular just to write docs for you.

If you aren't going to write them yourself and you're not willing to participate directly, you're being extremely unreasonable.

My suggestion? Stop typing replies and schedule a call with Andarist.

**Reply by with-heart** [reply]
Sorry, that might come across as more harsh than I meant it to be. It's just like you're the one with the Angular knowledge here, if you really want this addressed you're gonna have to help them help you ya know?

**Reply by Andarist** [reply]
It's worth noting that we only implemented React integration ourselves because that's what we know. I know the ins and outs of that ecosystem and I can **confidently** develop a reusable library for it. All other "integrations" (Vue, Svelte, and now Solid) were contributed by the community members - users of those frameworks. 

I can craft some implementation for Angular based on the docs but it probably wouldn't match its community standards. I feel like that would be a waste of time on my part because the Angular community wouldn't like to use such `@xstate/angular` that wouldn't match their patterns etc.

It's fairly easy to use XState in Angular. I just spent 10 minutes with the docs and got this: https://stackblitz.com/edit/angular-msmcak?file=src%2Fmain.ts . However, I still don't know how to wrap this in a reusable library that Angular devs would like to use. I need an actual Angular dev to tell me more about that ecosystem, its patterns etc

**Reply by stefumies** [reply]
@with-heart 

With that sort of reply it isn't particularly inviting to start to contribute if and when I have the time .

 That was particular misconstrued opinion of my discussions. I myself am not an Angular developer as such either, and I only had the XState documentation to go off , there there are links to implementations, I am / was not aware of the context as to the examples, and why they may have been implemented.

Seems right now , there is little impetus (for whatever reason to move forward. I would love to be able to contribute at some stage, but right now I have neither the time nor the depth of knowledge required) not (as incorrectly assumed) because I'm unwilling.


@Andarist 
Indeed, docs and deconstructed examples are how we have managed so far, this is why I came looking for a more integrated approach as it's currently time consuming. However, thank you very kindly for the example link, it looks initially like the simple example I'm seeking!

**Reply by with-heart** [reply]
Definitely not my best self in this thread. Sorry @stevematdavies!

**Comment by Avisa-GA** [comment]
Can you start by showing how you would do this without XState, and we can take that code and translate it to an XState example?

After this David, Would you please add the solution to xState Doc, So We can practice, also on YouTube channel will be fine too.
Thank you.


Sent from Yahoo Mail for iPhone


On Tuesday, January 24, 2023, 5:39 AM, David Khourshid ***@***.***> wrote:





Im approaching this merely as a Developer / User. I would like to see a very simple example (such as the Traffic light example) in Angular. Not with custom hacks and such but showing how to get and read the state in an Angular Service, via Rxjs or not, there need be no async calls, or anything.


Can you start by showing how you would do this without XState, and we can take that code and translate it to an XState example?

**Comment by niklas-wortmann** [comment]
Just jumping in this conversation. I was thinking of giving an angular implementation for xstate a try. I think this would be a valuable enhancement to the angular ecosystem. If there is any documentation around conventions for third party libraries, that would be very helpful

**Reply by Andarist** [reply]
> If there is any documentation around conventions for third party libraries, that would be very helpful

From the Angular's PoV or XState's? Most packages that integrate XState with UI frameworks have similar API - but we don't mind diverging from that. The primary focus of `@xstate/angular` should be to follow Angular conventions and not the conventions set for other UI frameworks by `@xstate/react`, `@xstate/vue`, etc.

**Reply by niklas-wortmann** [reply]
I was more thinking from an xstate perspective.  But I think we are good I will give it a try and report back here.

**Comment by jeremymwells** [comment]
Hi all. I was an early adopter of AngularJS, and have used Angular since. I want to see this exist and would like to help.

Angular uses decorators to add symbols to prototypes, which are interacted with at runtime. Compositionally, changing xstate dx to declare xstate/machine behaviors on TS classes would seem like the right approach to me. 

Here's a sketch of how something like this could take shape:

```
// machine composition -->

@Machine({
  initial: states.HomeState, // ctor
  states: [
    states.State1State, // ctor
    states.State2State, // ctor
    states.State3State, // ctor
    states.DoneState, // ctor
  ]
})
export class SampleMachine {
   /* machine level behaviors/etc here */
}

/***********************************************************/

// state composition -->
@State({
  initial: true,
  states: [
    states.State1aState, // ctor
    states.State1bState, // ctor
  ],
})
export class State1State {

  @Action({
    target: states.State1aState, // ctor
  })
  goTo1a(): void {
    /* transition logic here (etc) */
  }

}

/***********************************************************/

// angular provider in standalone component -->
@Component({
  selector: 'app-sample-machine',
  standalone: true,
  imports: [
    CommonModule
  ],
  providers: [
    MachineService.for([
      SampleMachine
    ])
  ],
  templateUrl: './sample-machine.component.html',
  styleUrls: ['./sample-machine.component.scss'],
})
export class SampleMachineComponent {
  machine = this._machineSvc.use(SampleMachine);

  constructor(
    private _machineSvc: MachineService // provided by @xstate/angular
  ) {  } 
  
  someUserAction() {
     this.machine.transition({ type: states.State1State });
  }
  
}
```
This is class-based, but I'd think the core logic in the decorators could be re-used for functional implementations.

Thoughts?

**Reply by jeremymwells** [reply]
The goal/idea is - at runtime, the `SampleMachine` prototype seen here would have a `Symbol('ɵmachine')` property on it that contains the same xstate machine definition we all know and love, and it would either be 1) accessible via the MachineService provider, or 2) handily abstracted by the MachineService
